### ISSUE TITLE
Engine benchmark regression on 2024-03-04

### ISSUE THREAD

**Issue by Akirathan** [enso-org/enso#9563]
We have some huge benchmark regressions on **2024-03-04**.

![image](https://github.com/enso-org/enso/assets/14013887/19e412ef-b432-4d62-90f9-15b192d1c2e7)

Regressed Engine benchmarks:
- [org_enso_interpreter_bench_benchmarks_semantic_IfVsCaseBenchmarks_caseBench3](https://enso-org.github.io/engine-benchmark-results/engine-benchs.html#org_enso_interpreter_bench_benchmarks_semantic_IfVsCaseBenchmarks_caseBench3)
  - **+577%**
- [org_enso_interpreter_bench_benchmarks_semantic_IfVsCaseBenchmarks_caseBench6](https://enso-org.github.io/engine-benchmark-results/engine-benchs.html#org_enso_interpreter_bench_benchmarks_semantic_IfVsCaseBenchmarks_caseBench6)
  - +378%
- [org_enso_interpreter_bench_benchmarks_semantic_IfVsCaseBenchmarks_ifBench3](https://enso-org.github.io/engine-benchmark-results/engine-benchs.html#org_enso_interpreter_bench_benchmarks_semantic_IfVsCaseBenchmarks_ifBench3)
  - +477% 
- [org_enso_interpreter_bench_benchmarks_semantic_IfVsCaseBenchmarks_ifBench6](https://enso-org.github.io/engine-benchmark-results/engine-benchs.html#org_enso_interpreter_bench_benchmarks_semantic_IfVsCaseBenchmarks_ifBench6)
  - +233%
- [org_enso_interpreter_bench_benchmarks_semantic_StringBenchmarks_lengthOfStrings](https://enso-org.github.io/engine-benchmark-results/engine-benchs.html#org_enso_interpreter_bench_benchmarks_semantic_StringBenchmarks_lengthOfStrings)
  - **+1400%**
-  [org_enso_interpreter_bench_benchmarks_semantic_WarningBenchmarks_noWarningsVecSum](https://enso-org.github.io/engine-benchmark-results/engine-benchs.html#org_enso_interpreter_bench_benchmarks_semantic_WarningBenchmarks_noWarningsVecSum)
    - **+2600%**
- [org_enso_interpreter_bench_benchmarks_semantic_WarningBenchmarks_randomElementsVecSum](https://enso-org.github.io/engine-benchmark-results/engine-benchs.html#org_enso_interpreter_bench_benchmarks_semantic_WarningBenchmarks_randomElementsVecSum)
  - +557%
- [org_enso_interpreter_bench_benchmarks_semantic_EqualsBenchmarks_equalsTrees](https://enso-org.github.io/engine-benchmark-results/engine-benchs.html#org_enso_interpreter_bench_benchmarks_semantic_EqualsBenchmarks_equalsTrees)
  - -82%

stdlib benchmarks appear to be fine.

Commits that likely caused the regressions:
- good: 39af372bcd4d93abfae94575e2c529ed92887049
- bad: 5d00a6110d5735c0116ddf869ec4f43433737a37
- between good and bad:  https://github.com/enso-org/enso/compare/39af372bcd4d93abfae94575e2c529ed92887049...5d00a6110d5735c0116ddf869ec4f43433737a37
- The one commit that is most likely responsible for the regression is #9190

**Comment by Akirathan** [comment]
`engine.BackgroundCompilation` was newly set to `false` in https://github.com/enso-org/enso/pull/9190/files#diff-7c0bf5c634ebfb606d229c275ac2cba4f74f2b156a6a674a43a6bc22d81968aaR58-R71. This is very bad idea for benchmarks. Let's revert this.

<blockquote><img src="https://avatars.githubusercontent.com/u/26887752?s=400&v=4" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> GitHub</div><div><strong><a href="https://github.com/enso-org/enso/pull/9190">Autoscoped constructors by JaroslavTulach · Pull Request #9190 · enso-org/enso</a></strong></div><div>Pull Request Description
Fixes #8645 by recognizing ~ prefix to constructor names.
Checklist
Please ensure that the following checklist has been satisfied before submitting the PR:

 The documentat...</div></blockquote>

**Comment by Akirathan** [comment]
@JaroslavTulach's theory is that `engine.BackgroundCompilation=false` should not affect performance. Let's ensure that this is the case (**DISCLAIMER: It is not - it causes huge regressions**).

Experimenting with:
```
sbt:runtime-benchmarks> run -wi 3 -w 10 -i 3 -r 10 org.enso.interpreter.bench.benchmarks.semantic.WarningBenchmarks.noWarningsVecSum
```
(noWarningsVecSum benchmark with 3x10 seconds warmup and 3x10 seconds measurement)

Local score with `--engine.BackgroundCompilation=false`: **0.183**
Local score with `--engine.BackgroundCompilation=true`: **0.009**

Setting `--engine.MultiTier` to either false or true does not seem to have any impact on the final score.

## Run with CompilationStatisticDetails=true
Running with `--engine.CompilationStatisticDetails` reveals that:

### BackgroundCompilation=false
Compilations: 25
Success: 24
Temporary bailouts: 0
Permanent bailouts: 1

![image](https://github.com/enso-org/enso/assets/14013887/9b3e3220-5845-4b40-befe-6234c7f7fad0)


### BackgroundCompilation=true
Compilations: 83
Success: 76
Temporary bailouts: 6
Permanent bailouts: 1

![image](https://github.com/enso-org/enso/assets/14013887/f4d7a102-e7d2-4b91-bf10-db20f2ad3caa)

This just tells us that there is a lot more compilation for `engine.BackgroundCompilation=true`, which is a bit obvious.


## TL;DR;
`--engine.BackgroundCompilation=false` causes huge regressions. We should not use it. Moreover, it is not used in production, so let's not use it for benchmarking. The same is true for `--engine.MultiTier=false`. In production, we use `--engine.MultiTier=true` (the default). Why use something different in benchmarks?

**Comment by JaroslavTulach** [comment]
### MultiTier

> In production, we use `--engine.MultiTier=true` (the default). Why use something different in benchmarks?

The IGV graphs are simpler to analyze with `MultiTier=false` and as you write...

> Setting --engine.MultiTier to either false or true does not seem to have any impact on the final score.

...it doesn't affect the result of final compilation. E.g. easier to disable multi tier compilation in benchmarks.
